pkg/remote/oci.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remote
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	imgname "github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/pkg/client/clientset/versioned/scheme"
+)
+
+// KeychainProvider is an input to the OCIResolver which returns a keychain for fetching remote images with
+// authentication.
+type KeychainProvider func() (authn.Keychain, error)
+
+// OCIResolver will attempt to fetch Tekton resources from an OCI compliant image repository.
+type OCIResolver struct {
```

Comment:
What's a typical usage of this type going to look like?  If it's most often going to be instantiated and immediately used like this:

```go
{
  myresolver := remote.OCIResolver{imageRef, kp}
  mytask := myresolver.GetTask(foo)
}
```

then I suggest simply exposing a `GetOCITask` func that takes `(imageReference string, keychainProvider KeychainProvider, taskName string)`.  WDYT?

Reply:
yeah that makes a ton of sense. I guess it comes down to how this evolves when we add a resolver for Git or other remote sources. I could see an argument for either route. 

I think the idea here is that it might be easier for a `reconciler` to use a factory to instantiate a `Resolver` and call it instead of having to know itself whether it should use an OCIResolver or a GitResolver or whatever else we add. We might even add a "local" resolver and embed the local task resolver into this format. Could be overkill or could be a nice abstraction. Thoughts?

Reply:
I can definitely see a factory fitting that use case.  I'm not sure it follows that we'd therefore need a `remote.Resolver` interface and creation of structs like this anywhere outside the factory's internals though. 

I think it'll ultimately depend on both the kind of configuration we expect to be available and the signature we want to expose for resolution. Examples of config might be that we allow specific resolvers to be turned on/off/configured via configmap / cli flags.  Examples of public signature might be a func like `ResolveTaskFromRef(taskRef)` that figures out the kind of resolution it needs to make without publicly exposing the Resolver type.

Not to say we need to commit to those kinds of design choices here, more that by introducing the Resolver machinery before we get to that stage we're already kind of leaning towards a design. And as often happens we might end up needing to unwind those kinds of choices based on work discovered as we dig into that stage more thoroughly.  And in turn that can lead to a some extra work reviewing in future as well.

None of this is a blocker at all from my POV, just raising it from my read-through. 👍 

/lgtm

Reply:
Yup. Makes a ton of sense. We can always wrap the resolver in a function like you mention to give a single interface to caller but I guess the next PR that embeds this in the reconciler will show if it makes it more or less "gross" and we can go from there 😄